ribs.py

```diff
@@ -123,6 +123,76 @@ def clear_screen(color):
     bottom_right = pg.display.get_surface().get_size()
     pg.draw.rect(window, color, (top_left, bottom_right))
 
+#
+# Simple physics and collision
+#
+
+def overlap_data(a, b):
+    """
+        Given that a and b overlap, extracts the shortest direction to move
+        along and the penetration.
+
+        returns -> normal, depth
+        (normal points from a)
+    """
+    # This code is based on SAT (the shortest distance is along one normal).
+    delta = a.centerx - b.centerx, a.centery - b.centery
+    span = (a.width + b.width) / 2, (a.height + b.height) / 2
+
+    # Pick the smallest overlapping axis
+    overlap = span[0] - abs(delta[0]), span[1] - abs(delta[1])
+    depth = min(overlap)
+
+    # Inline helper
+    sign = lambda x: 1 if x > 0 else -1
+
+    if abs(overlap[0]) < abs(overlap[1]):
+        normal = sign(delta[0]), 0
+    else:
+        normal = 0, sign(delta[1])
+
+    return normal, depth
+
+
+def solve_rect_overlap(a, b, vel_a=(0, 0), vel_b=(0, 0), mass_a=1, mass_b=1, bounce=0):
+    """
+        Solves the collision between a and b, with the mass and velocity as specified.
+
+        vel  - is the velocity of the body, a direction and a speed.
+        mass - is how much each body weighs, if set to 0 they cannot move.
+        bounce - is how bouncy the collision is. (For "correct" behaviour > 0.0, < 1.0)
+    """
```

Comment:
Maybe explain what "solving" in this case means? Most new people won't be used to the terminology.

```
Solve the collision between a and b, with the mass and velocity as specified.

Returns new values for a, b, vel_a and vel_b.
```